### PR TITLE
feat: add logger with userId and clientId

### PR DESCRIPTION
**Pull request by saleniuk**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

Currently logs are not associated with specific user and client which makes debugging harder.

### Solutions

Create a logger in user scope that can pass userId and clientId with each log.

#### Test Coverage (Optional)

- [ ] I have added automated test to this contribution

#### How to Test

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: logger/src/commonMain/kotlin/com/wire/kalium/logger/KaliumLogger.kt [hunk 1]
```diff
@@ -61,96 +62,143 @@ enum class KaliumLogLevel {
...
-    val severity = config.severity
+    fun logLevel(): KaliumLogLevel = config.logLevel()
 
     @Suppress("unused")
-    fun withFeatureId(featureId: ApplicationFlow): KaliumLogger {
-        val currentLogger = this
-        currentLogger.kermitLogger = kermitLogger.withTag("featureId:${featureId.name.lowercase()}")
-        return currentLogger
-    }
+    fun withFeatureId(featureId: ApplicationFlow): KaliumLogger = KaliumLogger(
```
Comment (vitorhugods) [inline logger/src/commonMain/kotlin/com/wire/kalium/logger/KaliumLogger.kt:109]: suggestion: maybe document this function to explain to the caller how it behaves when the tag includes a UserClientText already, etc.

File: logger/src/commonMain/kotlin/com/wire/kalium/logger/KaliumLogger.kt [hunk 2]
```diff
@@ -61,96 +62,143 @@ enum class KaliumLogLevel {
...
+        config = config,
+        tag = "featureId:${featureId.name.lowercase()}".let {
+            when (tag) {
+                is Tag.Text -> Tag.Text(it)
+                is Tag.UserClientText -> Tag.UserClientText(it, tag.data)
+            }
+        }
+    )
+
+    fun withUserDeviceData(data: () -> UserClientData): KaliumLogger = KaliumLogger(
```
Comment (vitorhugods) [inline logger/src/commonMain/kotlin/com/wire/kalium/logger/KaliumLogger.kt:119]: suggestion: also here. Document to make sure the caller understands the behaviour.

File: logger/src/commonMain/kotlin/com/wire/kalium/logger/KaliumLogger.kt [hunk 3]
```diff
@@ -61,96 +62,143 @@ enum class KaliumLogLevel {
...
-            KermitLogger(
-                config = StaticConfig(
-                    minSeverity = config.severityLevel, logWriters.asList()
-                ),
-                tag = config.tag
-            )
-        }
+class KaliumLogger(
+    private val config: Config = Config.DEFAULT,
+    private val tag: Tag = Tag.Text("KaliumLogger")
```
Comment (vitorhugods) [inline logger/src/commonMain/kotlin/com/wire/kalium/logger/KaliumLogger.kt:90]: suggestion: also here. Document what is this tag, how it behaves, etc.

File: logger/src/commonMain/kotlin/com/wire/kalium/logger/KaliumLogger.kt [hunk 4]
```diff
@@ -61,96 +62,143 @@ enum class KaliumLogLevel {
...
+                initialLevel = KaliumLogLevel.DISABLED,
+                initialLogWriterList = listOf(platformLogWriter()),
             )
         }
     }
 
+    sealed class Tag {
+        data class Text(val text: String) : Tag()
+        data class UserClientText(val prefix: String, val data: () -> UserClientData) : Tag()
+    }
```
Comment (vitorhugods) [inline logger/src/commonMain/kotlin/com/wire/kalium/logger/KaliumLogger.kt:180]: suggestion: a bit of Kdoc :)

File: logger/src/commonMain/kotlin/com/wire/kalium/logger/KaliumLogger.kt (hunk 1 above)
Comment (saleniuk) [reply logger/src/commonMain/kotlin/com/wire/kalium/logger/KaliumLogger.kt:116]: ✅

File: logger/src/commonMain/kotlin/com/wire/kalium/logger/KaliumLogger.kt (hunk 3 above)
Comment (saleniuk) [reply logger/src/commonMain/kotlin/com/wire/kalium/logger/KaliumLogger.kt:92]: ✅

File: logger/src/commonMain/kotlin/com/wire/kalium/logger/KaliumLogger.kt (hunk 4 above)
Comment (saleniuk) [reply logger/src/commonMain/kotlin/com/wire/kalium/logger/KaliumLogger.kt]: ✅

**Review by alexandreferris — APPROVED** [review]

**Review by gongracr — APPROVED** [review]

**Pull request merged**